File: Golang_src/medium/sumOfTwoIntegers_371.go
package medium

// Use Bit Manipulation via iterative:
// Time Complexity: O(logn)
// Space Complexity:O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 1.9 MB, less than 100.00%
func getSum(a int, b int) int {
	for b != 0 {
		carry := a & b << 1
		a = a ^ b
		b = carry
	}

	return a
}

// Use Bit Manipulation via recursive:
// Time Complexity: O(logn)
// Space Complexity:O(1)
// Runtime: 0 ms, faster than 100.00%
// Memory Usage: 2 MB, less than 31.94%
func getSum2(a int, b int) int {
	if b == 0 {
		return a
	}

	return getSum2((a ^ b), ((a & b) << 1))
}

File: Cpp_src/easy/sum_of_two_integers_371.cc
#include <iostream>
using namespace std;

// Use bit manipulation with iteration
// Time Complexity: O(logn)
// Space Complexity:O(1)
// Runtime: 4 ms, faster than 50.43%
// Memory Usage: 8.1 MB, less than 100.00%
class Solution
{
public:
    int getSum(int a, int b)
    {
        while (b)
        {
            int carry = ((a & b) & 0xffffffff) << 1;
            a ^= b;
            b = carry;
        }
        return a;
    }
};

// Use bit manipulation with recursive
// Time Complexity: O(logn)
// Space Complexity:O(1)
// Runtime: 4 ms, faster than 50.43%
// Memory Usage: 8.2 MB, less than 58.33%
class Solution2
{
public:
    int getSum(int a, int b)
    {
        return (b == 0) ? a : getSum((a ^ b), (((a & b) & (0xffffffff)) << 1));
    }
};